gcp_variant_transforms/libs/generate_report.py

```diff
@@ -0,0 +1,127 @@
+    info_conflicts = self._extract_conflicts(header_definitions.infos)
+    format_headers = resolved_header.formats
+    info_headers = resolved_header.infos
+    contents = list()
+    contents.extend(self._generate_contents(format_conflicts, format_headers))
+    contents.extend(self._generate_contents(info_conflicts, info_headers))
+    self._write_to_report(contents)
+
+  def _extract_conflicts(
+      self,
+      definitions  # type: Dict[str, Dict[Definition, List[str]]]
+      ):
+    # type: (...) -> Dict[str, Dict[Definition, List[str]]]
+    """Extracts the fields that have conflicted definitions."""
+    # len(v) > 1 means there are conflicted definitions for this field.
+    return dict([(k, v) for k, v in definitions.items() if len(v) > 1])
+
+  def _generate_contents(
+      self,
+      conflicts,  # type: Dict[str, Dict[Definition, List[str]]]
+      headers  # type: Dict[str, Dict[str, Union[str, int]]
+      ):
+    # type: (...) -> List[str]
+    """Generates the report contents.
+
+    The conflicted definitions, the file names and the resolutions are included
+    in the contents.
+    """
+    contents = list()
+    for field_id, definitions_to_files_map in conflicts.iteritems():
+      row = [
+          field_id,
+          self._extract_definitions_and_file_names(definitions_to_files_map),
+          self._extract_resolution(headers, field_id)
+      ]
+      contents.append(row)
+    return contents
+
+  def _extract_definitions_and_file_names(self, definition_to_files_map):
+    # type: (Dict[Definition, List[str]]) -> str
+    conflict_definitions = []
+    for definition, file_names in definition_to_files_map.iteritems():
+      definition = self._format_definition(definition.num, definition.type)
+      conflict_definitions.append(definition + ' in ' + str(file_names))
+    return ' '.join(conflict_definitions)
+
+  def _extract_resolution(self, header, filed_id):
+    # type: (Dict[str, Dict[str, Union[str, int]]], str) -> str
+    if filed_id not in header:
+      return self._NO_SOLUTION_MESSAGE
+    return self._format_definition(
+        header[filed_id][VcfParserHeaderKeyConstants.NUM],
+        header[filed_id][VcfParserHeaderKeyConstants.TYPE])
+
+  def _format_definition(self, num_value, type_value):
+    # type: (Union[str, int], str) -> str
+    formatted_definition = [
+        VcfParserHeaderKeyConstants.NUM + '=' + str(num_value),
+        VcfParserHeaderKeyConstants.TYPE + '=' + str(type_value)
+    ]
+    return ' '.join(formatted_definition)
+
+  def _write_to_report(self, contents):
+    # type: (List[str]) -> None
+    if self._directory and not os.path.exists(self._directory):
+      os.makedirs(self._directory)
+    file_path = os.path.join(self._directory, self._REPORT_NAME)
+    with open(file_path, 'w') as file_to_write:
+      writer = csv.writer(file_to_write)
+      if not contents:
+        writer.writerow(self._NO_CONFLICTS_MESSAGE)
+        return
```

Comment:
nit: I personally find using `else` to be more readable in this case as you only have one special case and the `else` block is short.